webui/src/App.js

```diff
@@ -21,86 +21,168 @@ import {
-let kCurValues = [new Array(num_panels).fill(0)];
-const max_size = 1000;
-let oldest = 0;
+// Returned `defaults` property will be undefined if the defaults are loading or reloading.
+// Call `reloadDefaults` to clear the defaults and reload from the server.
+function useDefaults() {
+  const [defaults, setDefaults] = useState();
 
-var ws;
-const wsCallbacks = {};
-const wsQueue = [];
+  const reloadDefaults = useCallback(() => setDefaults(undefined), [setDefaults]);
 
-function connect() {
-  ws = new WebSocket('ws://' + window.location.host + '/ws');
+  // Load defaults at mount and reload any time they are cleared.
+  useEffect(() => {
+    let cleaningUp = false;
+    let timeoutId = 0;
 
-  ws.addEventListener('open', function(ev) {
-    while (wsQueue.length > 0 && ws.readyState === 1) {
-      let msg = wsQueue.shift();
-      ws.send(JSON.stringify(msg));
+    const getDefaults = () => {
+      clearTimeout(timeoutId);
+      fetch('/defaults').then(res => res.json()).then(data => {
+        if (!cleaningUp) {
+          setDefaults(data);
+        }
+      }).catch(reason => {
+        if (!cleaningUp) {
+          timeoutId = setTimeout(getDefaults, 1000);
+        }
+      });
     }
-  });
 
-  ws.addEventListener('error', function(ev) {
-    ws.close();
-  });
+    if (!defaults) {
+      getDefaults();
+    }
+
+    return () => {
+      cleaningUp = true;
+      clearTimeout(timeoutId);
+    };
+  }, [defaults]);
+
+  return { defaults, reloadDefaults };
+}
 
-  ws.addEventListener('close', function(ev) {
-    setTimeout(connect, 1000);
+// returns { curValuesRef, emit, isWsReady, wsCallbacksRef }
+// curValuesRef tracks values as well as thresholds.
+// Use emit to send events to the server.
+// isWsReady indicates that the websocket is connected and has received
+// its first values reading from the server.
+// onCloseWs is called when the websocket closes for any reason,
+// unless the hooks is already doing effect cleanup.
+// The expectation is that onCloseWs is used to reload default thresholds
+// and provide new ones to trigger a new connection.
+function useWsConnection({ defaults, onCloseWs }) {
+  const [isWsReady, setIsWsReady] = useState(false);
+
+  const curValuesRef = useRef({
+
+    // A history of the past 'max_size' values fetched from the backend.
+    // Used for plotting and displaying live values.
+    // We use a cyclical array to save memory.
+    kCurValues: [],
+    oldest: 0,
+
+    // Keep track of the current thresholds fetched from the backend.
+    kCurThresholds: [],
   });
 
-  ws.addEventListener('message', function(ev) {
-    const data = JSON.parse(ev.data)
-    const action = data[0];
-    const msg = data[1];
+  const wsRef = useRef();
+  const wsCallbacksRef = useRef({});
 
-    if (wsCallbacks[action]) {
-      wsCallbacks[action](msg);
+  const emit = useCallback((msg) => {
+    // App should wait for isWsReady to send messages.
+    if (!wsRef.current || !isWsReady) {
```

Comment:
Do we need `isWsReady` variable? Does the the `wsRef.current.readyState` field suffice?

Reply:
I did it this way for two reasons. One is because I don't actually mark it ready until values start coming in, but I'm changing that back as you requested. The other reason is to make sure `App` re-renders when it changes. Because `isWsReady` is React state, calling `setIsWsReady` will cause the component that called `useState` to re-render. It's less about this particular check in `emit` but I thought it made sense to check the same thing as App.